kyo-core/shared/src/main/scala/kyo/env.scala

```diff
@@ -0,0 +1,47 @@
+package kyo
+
+import kyo.Tag.Intersection
+import scala.collection.immutable.HashMap
+
+opaque type Env[+R] = HashMap[Tag[?], Any]
+extension [R](self: Env[R])
+
+    private def fatal(using t: Tag[?]): Nothing =
+        throw new RuntimeException(s"fatal: kyo.Env of contents [$self] missing value of type: [${t.show}].")
+
+    def get[A >: R](using Tag[A]): A =
+        getOrElse(fatal)
+
+    def getOrElse[A >: R](default: => A)(using t: Tag[A]): A =
+        val result = self.getOrElse(t, null)
+        if isNull(result) then
+            default
+        else
+            result.asInstanceOf[A]
+        end if
+    end getOrElse
+
+    def add[A: Tag](a: A): Env[R & A] =
+        self.updated(Tag[A], a)
```

Comment:
Can you receive the `Tag` directly as an implicit parameter instead? I'm not sure if there'd be a noticeable performance difference here but avoiding the macro indirection should reduce the size of the generated bytecode.

Reply:
Sure, will do.